File: class18/java/TCC1801-FirstFollow.java
// TCC1801-FirstFollow.java | Aula 18
// Calcula conjuntos First e Follow de uma GLC.
// Espelho Java do TCC1801-FirstFollow.py (mesma gramática e saídas).
import java.util.*;

public class TCC1801FirstFollow {

    // Gramática: não-terminal → lista de produções
    static Map<String, List<List<String>>> gramatica = new LinkedHashMap<>();
    static Set<String> naoTerminais = new LinkedHashSet<>();
    static Set<String> terminais = new LinkedHashSet<>();

    static {
        gramatica.put("E",  Arrays.asList(Arrays.asList("T", "E'")));
        gramatica.put("E'", Arrays.asList(Arrays.asList("+", "T", "E'"), Arrays.asList("ε")));
        gramatica.put("T",  Arrays.asList(Arrays.asList("F", "T'")));
        gramatica.put("T'", Arrays.asList(Arrays.asList("*", "F", "T'"), Arrays.asList("ε")));
        gramatica.put("F",  Arrays.asList(Arrays.asList("(", "E", ")"), Arrays.asList("id")));

        naoTerminais.addAll(gramatica.keySet());
        for (List<List<String>> prods : gramatica.values()) {
            for (List<String> prod : prods) {
                for (String s : prod) {
                    if (!naoTerminais.contains(s)) terminais.add(s);
                }
            }
        }
        terminais.remove("ε");
    }

    static Map<String, Set<String>> first = new LinkedHashMap<>();
    static Map<String, Set<String>> follow = new LinkedHashMap<>();

    static Set<String> firstSimbolo(String s, Set<String> visitados) {
        if (terminais.contains(s) || s.equals("ε")) return new LinkedHashSet<>(Arrays.asList(s));
        if (first.containsKey(s)) return first.get(s);
        if (visitados.contains(s)) return new LinkedHashSet<>();
        visitados.add(s);
        Set<String> resultado = new LinkedHashSet<>();
        for (List<String> prod : gramatica.get(s)) {
            for (String simb : prod) {
                Set<String> f = firstSimbolo(simb, visitados);
                resultado.addAll(f); resultado.remove("ε");
                if (!f.contains("ε")) break;
            }
            // se todos derivam ε
            boolean todosEpsilon = true;
            for (String simb : prod) {
                Set<String> f = firstSimbolo(simb, visitados);
                if (!f.contains("ε")) { todosEpsilon = false; break; }
            }
            if (todosEpsilon) resultado.add("ε");
        }
        first.put(s, resultado);
        return resultado;
    }

    static Set<String> firstDeSequencia(List<String> seq) {
        Set<String> resultado = new LinkedHashSet<>();
        for (String s : seq) {
            Set<String> f = first.containsKey(s) ? first.get(s) : new LinkedHashSet<>(Arrays.asList(s));
            resultado.addAll(f); resultado.remove("ε");
            if (!f.contains("ε")) break;
        }
        boolean todosEpsilon = true;
        for (String s : seq) {
            Set<String> f = first.containsKey(s) ? first.get(s) : new LinkedHashSet<>(Arrays.asList(s));
            if (!f.contains("ε")) { todosEpsilon = false; break; }
        }
        if (todosEpsilon) resultado.add("ε");
        return resultado;
    }

    static void calcularFirst() {
        for (String nt : naoTerminais) firstSimbolo(nt, new HashSet<>());
    }

    static void calcularFollow() {
        for (String nt : naoTerminais) follow.put(nt, new LinkedHashSet<>());
        follow.get("E").add("$");
        boolean mudou = true;
        while (mudou) {
            mudou = false;
            for (String A : naoTerminais) {
                for (List<String> prod : gramatica.get(A)) {
                    for (int i = 0; i < prod.size(); i++) {
                        String B = prod.get(i);
                        if (!naoTerminais.contains(B)) continue;
                        List<String> resto = prod.subList(i + 1, prod.size());
                        Set<String> fResto = resto.isEmpty() ? new LinkedHashSet<>(Arrays.asList("ε")) : firstDeSequencia(resto);
                        Set<String> novo = new LinkedHashSet<>(follow.get(B));
                        novo.addAll(fResto); novo.remove("ε");
                        if (fResto.contains("ε") || resto.isEmpty()) {
                            novo.addAll(follow.get(A));
                        }
                        if (!novo.equals(follow.get(B))) {
                            follow.put(B, novo);
                            mudou = true;
                        }
                    }
                }
            }
        }
    }

    public static void main(String[] args) {
        calcularFirst();
        calcularFollow();

        System.out.println("=== FIRST ===");
        for (String nt : naoTerminais) {
            System.out.println("  First(" + nt + ") = " + first.get(nt));
        }

        System.out.println("\n=== FOLLOW ===");
        for (String nt : naoTerminais) {
            System.out.println("  Follow(" + nt + ") = " + follow.get(nt));
        }
    }
}

File: class18/python/TCC1801-FirstFollow.py
#!/usr/bin/env python3
"""TCC1801-FirstFollow.py | Aula 18
Calcula conjuntos First e Follow de uma GLC.
Gramática exemplo:
  E  → T E'
  E' → + T E' | ε
  T  → F T'
  T' → * F T' | ε
  F  → ( E ) | id
"""
from collections import defaultdict


# Gramática: não-terminal → lista de produções (cada produção é lista de símbolos)
gramatica = {
    "E":  [["T", "E'"]],
    "E'": [["+", "T", "E'"], ["ε"]],
    "T":  [["F", "T'"]],
    "T'": [["*", "F", "T'"], ["ε"]],
    "F":  [["(", "E", ")"], ["id"]],
}

nao_terminais = set(gramatica.keys())
terminais = set()
for prods in gramatica.values():
    for prod in prods:
        for s in prod:
            if s not in nao_terminais:
                terminais.add(s)
terminais.discard("ε")


def first_simbolo(s: str, first: dict, visitados: set) -> set:
    if s in terminais or s == "ε":
        return {s}
    if s in first:
        return first[s]
    if s in visitados:
        return set()
    visitados.add(s)
    resultado = set()
    for prod in gramatica[s]:
        for simb in prod:
            f = first_simbolo(simb, first, visitados)
            resultado |= f - {"ε"}
            if "ε" not in f:
                break
        else:
            resultado.add("ε")
    first[s] = resultado
    return resultado


def calcular_first() -> dict:
    first = {}
    for nt in nao_terminais:
        first_simbolo(nt, first, set())
    return first


def first_de_sequencia(seq: list, first: dict) -> set:
    resultado = set()
    for s in seq:
        f = first.get(s, {s})
        resultado |= f - {"ε"}
        if "ε" not in f:
            break
    else:
        resultado.add("ε")
    return resultado


def calcular_follow(first: dict) -> dict:
    follow = defaultdict(set)
    follow["E"].add("$")  # símbolo inicial
    mudou = True
    while mudou:
        mudou = False
        for A in nao_terminais:
            for prod in gramatica[A]:
                for i, B in enumerate(prod):
                    if B not in nao_terminais:
                        continue
                    resto = prod[i + 1:]
                    f_resto = first_de_sequencia(resto, first) if resto else {"ε"}
                    novo = follow[B] | (f_resto - {"ε"})
                    if "ε" in f_resto or not resto:
                        novo |= follow[A]
                    if novo != follow[B]:
                        follow[B] = novo
                        mudou = True
    return follow


if __name__ == "__main__":
    first = calcular_first()
    follow = calcular_follow(first)

    print("=== FIRST ===")
    for nt in sorted(nao_terminais):
        print(f"  First({nt}) = {sorted(first[nt])}")

    print("\n=== FOLLOW ===")
    for nt in sorted(nao_terminais):
        print(f"  Follow({nt}) = {sorted(follow[nt])}")

    print("\n=== Tabela LL(1) (parcial) ===")
    print(f"{'':>6}", end="")
    for t in sorted(terminais | {"$"}):
        print(f"{t:>12}", end="")
    print()
    for nt in sorted(nao_terminais):
        print(f"{nt:>6}", end="")
        for t in sorted(terminais | {"$"}):
            celula = ""
            for prod in gramatica[nt]:
                f = first_de_sequencia(prod, first)
                if t in f:
                    celula = " ".join(prod)
                    break
                if "ε" in f and t in follow[nt]:
                    celula = " ".join(prod)
                    break
            print(f"{celula:>12}", end="")
        print()
